hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancer.java

```diff
@@ -0,0 +1,213 @@
+    mockNodeManager = new MockNodeManager(nodesInCluster);
+    containerBalancer = new ContainerBalancer(mockNodeManager, containerManager,
+        replicationManager, conf);
+  }
+
+  /**
+   * Checks whether ContainerBalancer is correctly updating the list of source
+   * nodes with varying values of Threshold.
+   */
+  @Test
+  public void initializeIterationShouldUpdateSourceNodesWhenThresholdChanges() {
+    List<DatanodeUsageInfo> expectedSourceNodes;
+    List<DatanodeUsageInfo> sourceNodesAccordingToBalancer;
+
+    // check for random threshold values
+    for (int i = 0; i < 50; i++) {
+      double randomThreshold = Math.random();
+
+      balancerConfiguration.setThreshold(String.valueOf(randomThreshold));
+      containerBalancer.start(balancerConfiguration);
+      expectedSourceNodes = determineExpectedSourceNodes(randomThreshold);
+      sourceNodesAccordingToBalancer = containerBalancer.getSourceNodes();
+
+      Assert.assertEquals(
+          expectedSourceNodes.size(), sourceNodesAccordingToBalancer.size());
+
+      for (int j = 0; j < expectedSourceNodes.size(); j++) {
+        Assert.assertEquals(expectedSourceNodes.get(j).getDatanodeDetails(),
+            sourceNodesAccordingToBalancer.get(j).getDatanodeDetails());
+      }
+      containerBalancer.stop();
+    }
+
+  }
+
+  /**
+   * Checks whether the list of source is empty when the cluster is balanced.
+   */
+  @Test
+  public void sourceNodesListShouldBeEmptyWhenClusterIsBalanced() {
+    balancerConfiguration.setThreshold("0.99");
+    containerBalancer.start(balancerConfiguration);
+
+    Assert.assertEquals(0, containerBalancer.getSourceNodes().size());
+    containerBalancer.stop();
+  }
+
+  /**
+   * Checks whether ContainerBalancer stops when the limit of
+   * MaxDatanodesToBalance is reached.
+   */
+  @Test
+  public void containerBalancerShouldStopWhenMaxDatanodesToBalanceIsReached() {
+    balancerConfiguration.setMaxDatanodesToBalance(2);
+    balancerConfiguration.setThreshold("0");
+    containerBalancer.start(balancerConfiguration);
+
+    Assert.assertFalse(containerBalancer.isBalancerRunning());
+    containerBalancer.stop();
+  }
+
+  /**
+   * Determines source nodes, that is, over and under utilized nodes,
+   * according to the generated utilization values for nodes and the threshold.
+   *
+   * @param threshold A fraction from range 0 to 1.
+   * @return List of DatanodeUsageInfo containing the expected(correct)
+   * source nodes.
+   */
+  private List<DatanodeUsageInfo> determineExpectedSourceNodes(
+      double threshold) {
+    double lowerLimit = averageUtilization - threshold;
+    double upperLimit = averageUtilization + threshold;
+
+    // use node utilizations to determine over and under utilized nodes
+    List<DatanodeUsageInfo> expectedSourceNodes = new ArrayList<>();
+    for (int i = 0; i < numberOfNodes; i++) {
+      if (nodeUtilizations.get(numberOfNodes - i - 1) > upperLimit) {
```

Comment:
Can we use nodesInCluster directly to determine the source nodes?

Reply:
By using a separate method, varying thresholds can be passed as argument to determine the expected list for each threshold.